### PR TITLE
Feature/admin portal route pages

### PR DESCRIPTION
**Pull request by Gubbsy**
## Proposed changes

Creation of all the pages and functionality to facilitate administrative main contact and organisation changes through a portal that is only accessible to users with the `Admin` role. 

This includes all the required routing, controllers, and context to allow the User, Organisation, Funding Application and Pre-Application database entities to be updated through a simplified UI upon applicant request. 

A new audit table has also been added to track the use of the Admin portal (and potentially later expanded to the wider service). The audit table records:
- the entities changed (record type and id),
- the values before and after the change, 
- the Admin who made the change

### CODE REVIEW

File: app/controllers/admin_portal/application_to_move_controller.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+# Todo revisit for large.  This could look at the award type.
```
Comment (ptrelease) [inline app/controllers/admin_portal/application_to_move_controller.rb:1]: Code for large applications addressed - so I will remove this.

File: app/controllers/admin_portal/update_contact_details_controller.rb [hunk 2]
```diff
@@ -0,0 +1,196 @@
+# TODO - Audit
```
Comment (ptrelease) [inline app/controllers/admin_portal/update_contact_details_controller.rb:1]: I will write something here

File: app/controllers/admin_portal/update_contact_details_controller.rb [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+  NOT_FOUND_ERR = 'Salesforce Contact Id not found'
+
+  INCOMPLETE_ERR = "Contact details are incomplete in Salesforce. " \
+    "Check Last Name, Mailing Street, " \
+      "Mailing City and Mailing Zip/Postal Code"
+
+  def show
+    retrieve_user
+
```
Comment (ptrelease) [inline app/controllers/admin_portal/update_contact_details_controller.rb:18]: will tidy

File: app/controllers/admin_portal/update_contact_details_controller.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    @apps_to_move = @user_details.organisations.first.funding_applications
+    @preapps_to_move = @user_details.organisations.first.pre_applications
+  end
+
+  # Gets a Contact details from Salesforce.
+  # Handle deleted or invalid contact IDs
+  #
+  # @param [String] salesforce_contact_id
+  # @return [RestforceResponse] contact_from_salesforce
+  def get_contact_from_salesforce(salesforce_contact_id)
```
Comment (ptrelease) [inline app/controllers/admin_portal/update_contact_details_controller.rb:63]: I will fix the rdoc here

File: app/controllers/concerns/admin_portal_context.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+  end
+
+  def get_contact_salesforce_link(salesforce_contact_id)
+    return SALESFORCE_URL_BASE + "Contact/#{salesforce_contact_id}/view"
+  end
+
+  def get_salesforce_organisation_link(salesforce_account_id)
+    return SALESFORCE_URL_BASE + "Account/#{salesforce_account_id}/view"
```
Comment (ptrelease) [inline app/controllers/concerns/admin_portal_context.rb:37]: Test.  Does the SALESFORCE_URL_BASE need need a trailing slash to work?  If so, interpolate the env var with the main string - as double slashes handled better than none at all.

File: app/controllers/concerns/auditor.rb [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    audit_row.action_successful = action_successful
+
+    audit_row.record_changes = changes_hash
+
+    audit_row.request_id = request_id
+
+    audit_row.redact_date = redact_date
+
+    audit_row.save!
+
```
Comment (ptrelease) [inline app/controllers/concerns/auditor.rb:51]: Add logging

File: app/controllers/concerns/funding_application_context.rb [hunk 7]
```diff
@@ -104,7 +104,8 @@ def invalid_view_for_progress_and_spend?
         @funding_application&.del_250k_to_5mm?  ||
           @funding_application&.is_100_to_250k? ||
             @funding_application&.is_10_to_100k? ||
-              @funding_application&.dev_to_100k?
+              @funding_application&.dev_to_100k? ||
+                @funding_application&.migrated_large_delivery?
```
Comment (ptrelease) [inline app/controllers/concerns/funding_application_context.rb:109]: Why isn't this in main branch already

File: app/controllers/dashboard_controller.rb [hunk 8]
```diff
@@ -47,6 +48,11 @@ def show
             if funding_application.open_medium.present? && \
               awarded(funding_application, @salesforce_api_instance)
 
+            @migrated_large_delivery_grants_for_payment.push(
+              funding_application
+            ) if funding_application.migrated_large_delivery? &&
+              funding_application.payment_can_start?
+
```
Comment (ptrelease) [inline app/controllers/dashboard_controller.rb:55]: Why isn't this in main branch already

File: app/controllers/dashboard_controller.rb [hunk 9]
```diff
@@ -277,6 +292,7 @@ def get_large_payments_path(funding_application)
     end
 
   end
+  helper_method :get_large_payments_path
```
Comment (ptrelease) [inline app/controllers/dashboard_controller.rb:295]: Loads of code here that I would expect to be in main already.

File: app/controllers/concerns/admin_portal_context.rb (hunk 5 above)
Comment (ptrelease) [reply app/controllers/concerns/admin_portal_context.rb:37]: Initalised URl from SF connection on app startup

File: config/initializers/constants.rb [hunk 10]
```diff
@@ -0,0 +1,3 @@
+include SalesforceApi
+
+SALESFORCE_URL_BASE = SalesforceApiClient.new.get_salesforce_url
```
Comment (Gubbsy) [inline config/initializers/constants.rb:3]: No newline

File: app/controllers/concerns/admin_portal_context.rb [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+  # This method retrieves a SfxPtsPayment instance based on the id
+  # found in the URL parameters and the current authenticated
+  # user's id, setting it as an instance variable for use in
+  # related controllers.
+  #
+  # If no SfxPtsPayment instance matching the parameters is found,
+  # or the SfxPtsPayment instance is not suitable for the route being used,
+  # then the user is redirected to the applications dashboard.
+	def ensure_user_is_admin
```
Comment (Gubbsy) [inline app/controllers/concerns/admin_portal_context.rb:16]: Strange indenting of method dec

File: app/controllers/concerns/auditor.rb [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+end
```
Comment (Gubbsy) [inline app/controllers/concerns/auditor.rb:54]: No new line

File: app/helpers/admin_portal_helper.rb [hunk 13]
```diff
@@ -0,0 +1,430 @@
...
+            id: "Application not found",
+            ref_no:  "Application not found",
+            type: UNKNOWN,
+            title: "Application not found",
+            salesforce_id: "Application not found"
+          }
+
+  end
+
+end
```
Comment (Gubbsy) [inline app/helpers/admin_portal_helper.rb:430]: No new line

File: app/models/audit.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+class Audit < ApplicationRecord
+
+  enum audit_action: {
+    admin_contact_change: 0,
+    admin_organisation_change: 1,
+    admin_application_change: 2,
+  }
+
+end
```
Comment (Gubbsy) [inline app/models/audit.rb:9]: No new line

File: config/initializers/constants.rb (hunk 10 above)
Comment (ptrelease) [reply config/initializers/constants.rb:3]: Added

File: app/controllers/concerns/admin_portal_context.rb (hunk 11 above)
Comment (ptrelease) [reply app/controllers/concerns/admin_portal_context.rb:16]: Looks like git formatting issue.  Fine in code.[](url)

File: app/controllers/concerns/auditor.rb (hunk 12 above)
Comment (ptrelease) [reply app/controllers/concerns/auditor.rb:54]: added

File: app/models/audit.rb (hunk 14 above)
Comment (ptrelease) [reply app/models/audit.rb:9]: Added

File: app/controllers/concerns/admin_portal_context.rb [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+      "/r/Contact/"\
+        "#{salesforce_contact_id}"\
+          "/view"
+  end
+
+  # Returns a link that can be used to find the Account in Salesforce
+  # @return [String] url Link to the Account in SF
+  def get_salesforce_organisation_link(
+    salesforce_account_id,
+    client = SalesforceApiClient.new
```
Comment (ptrelease) [inline app/controllers/concerns/admin_portal_context.rb:43]: Pt remove line 43, - not used

**Review by ptrelease — APPROVED** [review]

**Pull request merged**